### PR TITLE
feat: extend channel priming logic to also send fake requests

### PR DESCRIPTION
**Pull request by igorbernstein2**
This PR upgrades the existing channel refresher to send a fake read row request to ensure that server side caches are warmed up. The new refresher will create a temporary stub around the new channel, which will reuse all of the existing request decorations (resource headers and user agent)

### CODE REVIEW

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStub.java [hunk 1]
```diff
@@ -120,65 +123,89 @@
...
-      ClientContext clientContext,
-      Tagger tagger,
-      StatsRecorder statsRecorder) {
-    this.settings = settings;
+  public static EnhancedBigtableStubSettings finalizeSettings(
+      EnhancedBigtableStubSettings settings, Tagger tagger, StatsRecorder stats)
+      throws IOException {
+    EnhancedBigtableStubSettings.Builder builder = settings.toBuilder();
+
+    if (settings.isRefreshingChannel()) {
```
Comment (kolea2) [inline google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStub.java:136]: Can we add some tests around this to ensure the proper values are set in the settings?

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStubSettings.java [hunk 2]
```diff
@@ -760,17 +787,10 @@ public EnhancedBigtableStubSettings build() {
       Preconditions.checkState(projectId != null, "Project id must be set");
       Preconditions.checkState(instanceId != null, "Instance id must be set");
 
-      // Set ChannelPrimer on TransportChannelProvider so channels will gracefully refresh
-      // connections to Cloud Bigtable service
       if (isRefreshingChannel) {
         Preconditions.checkArgument(
             getTransportChannelProvider() instanceof InstantiatingGrpcChannelProvider,
```
Comment (kolea2) [inline google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStubSettings.java:792]: should this check be moved to where we are setting the transport provider, https://github.com/googleapis/java-bigtable/pull/398/files#diff-dad9668e639c962562540f7f70a24138R144?

File: google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/stub/BigtableChannelPrimerTest.java [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+  Server server;
+  FakeService fakeService;
+  MetadataInterceptor metadataInterceptor;
+  BigtableChannelPrimer primer;
+  ManagedChannel channel;
+  private LogHandler logHandler;
+
+  @Before
+  public void setup() throws IOException {
+    try (ServerSocket ss = new ServerSocket(0)) {
```
Comment (kolea2) [inline google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/stub/BigtableChannelPrimerTest.java:68]: can we abstract any of this out into test helpers?

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStub.java [hunk 4]
```diff
@@ -120,65 +123,89 @@
...
   }
 
-  @InternalApi("Visible for testing")
-  public EnhancedBigtableStub(
-      EnhancedBigtableStubSettings settings,
-      ClientContext clientContext,
-      Tagger tagger,
-      StatsRecorder statsRecorder) {
-    this.settings = settings;
+  public static EnhancedBigtableStubSettings finalizeSettings(
```
Comment (kolea2) [inline google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStub.java:131]: this is a little confusing to me. Why did we separate these pieces out?

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStub.java [hunk 5]
```diff
@@ -120,65 +123,89 @@
...
+      builder.setCredentialsProvider(FixedCredentialsProvider.create(credentials));
+
+      // Inject the primer
+      InstantiatingGrpcChannelProvider transportProvider =
+          (InstantiatingGrpcChannelProvider) settings.getTransportChannelProvider();
+
+      builder.setTransportChannelProvider(
+          transportProvider
+              .toBuilder()
+              .setChannelPrimer(
```
Comment (tonytanger) [inline google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStub.java:151]: It would be nice to have a comment here explaining the consequence of "setChannelPrimer" that the channel is created once, but periodically creates a new channel, runs BigtableChannelPrimer on the newly created channel and replaces the existing one with the new one.

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStubSettings.java (hunk 2 above)
Comment (igorbernstein2) [reply google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStubSettings.java:792]: No, I want the validation logic to be as close to usage as possible.

File: google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/stub/BigtableChannelPrimerTest.java (hunk 3 above)
Comment (igorbernstein2) [reply google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/stub/BigtableChannelPrimerTest.java:68]: Lets track this separately, it will need a bit of refactoring in the other tests

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStub.java (hunk 4 above)
Comment (igorbernstein2) [reply google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStub.java:131]: Not sure which specific pieces you are referring to, but I'll try to give the background to the current structure.

There a couple of phases of client construction:
1. user uses the builder to configure their settings. At this point the user owns the object so modifying it in place during build() is very surprising
2. stub creation: at this point the user passed us the settings object. This is the correct place to augment the settings as the changes won't be visible in the user object

Previously step 2 was handled when we created ClientContext, which is a manifestation of the user settings.  Which is the correct place to do this. Unfortunately for this new feature, the primer needs to be injected before the ClientContext is created (because ClientContext instantiates the grpc connections). So the settings need to be augmented before the ClientContext is created. So ClientContext.create(settings).toBuilder().augment() was changed to ClientContext.create(settings.toBuilder().augment()). 

At this point the logic of patching settings got pretty hefty and fairly distinct from callable chain creation, so it was extracted to a helper method.

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/BigtableChannelPrimer.java [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+  private void primeChannelUnsafe(ManagedChannel managedChannel) throws IOException {
+    if (tableIds.isEmpty()) {
+      waitForChannelReady(managedChannel);
+    } else {
+      sendPrimeRequests(managedChannel);
+    }
+  }
+
+  private void waitForChannelReady(ManagedChannel managedChannel) {
+    for (int i = 0; i < 10; i++) {
```
[the remark this answers is not in the document]
Comment (igorbernstein2) [reply google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/BigtableChannelPrimer.java]: done

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/BigtableChannelPrimer.java [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+  public void primeChannel(ManagedChannel managedChannel) {
+    try {
+      primeChannelUnsafe(managedChannel);
+    } catch (IOException | RuntimeException e) {
+      LOG.warning(
+          String.format("Unexpected error while trying to prime a channel: %s", e.getMessage()));
+    }
+  }
+
+  private void primeChannelUnsafe(ManagedChannel managedChannel) throws IOException {
```
[the remark this answers is not in the document]
Comment (igorbernstein2) [reply google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/BigtableChannelPrimer.java:110]: Because it could throw an exception, does the priming api expect to handle the exceptions?

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStub.java [hunk 8]
```diff
@@ -120,65 +123,89 @@
...
-                                    TagValue.create(settings.getInstanceId()))
-                                .put(
-                                    RpcMeasureConstants.BIGTABLE_APP_PROFILE_ID,
-                                    TagValue.create(settings.getAppProfileId()))
-                                .build()),
-                        // Add user configured tracer
-                        clientContext.getTracerFactory())))
-            .build();
+    // Inject Opencensus instrumentation
+    builder.setTracerFactory(
```
[the remark this answers is not in the document]
Comment (igorbernstein2) [reply google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStub.java:162]: Lets merge this with the discussion above "this is a little confusing to me. Why did we separate these pieces out?"

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStub.java (hunk 5 above)
Comment (igorbernstein2) [reply google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStub.java:151]: I dont quite follow, can you rephrase?

File: google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/stub/BigtableChannelPrimerTest.java (hunk 3 above)
Comment (kolea2) [reply google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/stub/BigtableChannelPrimerTest.java:68]: sgtm, can you please file an issue?

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStub.java (hunk 4 above)
Comment (kolea2) [reply google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStub.java:131]: Thanks! By 'pieces' I was mostly referring to the channel priming and the tracing - wasn't quite sure how they went together. Do you think it would be clearer if we had a `finalizeSettings` which delegates to each sub instantiation (channel priming, tracing)?

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/BigtableChannelPrimer.java [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+  }
+
+  @Override
+  public void primeChannel(ManagedChannel managedChannel) {
+    try {
+      primeChannelUnsafe(managedChannel);
+    } catch (IOException | RuntimeException e) {
+      LOG.warning(
+          String.format("Unexpected error while trying to prime a channel: %s", e.getMessage()));
+    }
```
Comment (kolea2) [inline google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/BigtableChannelPrimer.java:107]: (see codecov warning) is this really not covered?

File: google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/stub/BigtableChannelPrimerTest.java (hunk 3 above)
Comment (igorbernstein2) [reply google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/stub/BigtableChannelPrimerTest.java:68]: Done https://github.com/googleapis/java-bigtable/issues/403

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStub.java (hunk 4 above)
Comment (igorbernstein2) [reply google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStub.java:131]: I'd like to leave it as is, we can factor those out when we get a 3rd feature

Comment (kolea2) [reply google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStub.java:131]: ok, then for now I think a comment would be helpful

Comment (igorbernstein2) [reply google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStub.java:131]: done

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/BigtableChannelPrimer.java (hunk 9 above)
Comment (igorbernstein2) [reply google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/BigtableChannelPrimer.java:107]: That branch is for unexpected errors, I added a test to handle that as well

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStub.java (hunk 1 above)
Comment (igorbernstein2) [reply google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStub.java:140]: I added a test for testChannelPrimerConfigured, which should cover the behavior of it
